cpp/src/join/full_join.cuh

```diff
@@ -0,0 +1,194 @@
+        size_type * const buffer_capacity,
+        const size_type buffer_size,
+        const size_type expand_size,
+        cudaStream_t stream) {
+    size_type requested_size = buffer_size + expand_size;
+    //No need to proceed if the buffer can contain requested additional elements
+    if (*buffer_capacity >= requested_size) {
+        return GDF_SUCCESS;
+    }
+    data_type * new_buffer{nullptr};
+    data_type * old_buffer = *buffer;
+    RMM_TRY( RMM_ALLOC((void**)&new_buffer, requested_size*sizeof(data_type), stream) );
+    CUDA_TRY( cudaMemcpy(new_buffer, old_buffer, buffer_size*sizeof(data_type), cudaMemcpyDeviceToDevice) );
+    RMM_TRY( RMM_FREE(old_buffer, stream) );
+    *buffer = new_buffer;
+    *buffer_capacity = requested_size;
+
+    return GDF_SUCCESS;
+}
+
+/* --------------------------------------------------------------------------*/
+/**
+* @Synopsis  Adds indices that are missing in r_index_ptr at the ends and places
+* JoinNoneValue to the corresponding l_index_ptr.
+*
+* @Param l_index_ptr Address of the left indices
+* @Param r_index_ptr Address of the right indices
+* @Param index_capacity Amount of memory allocated for left and right indices
+* @Param index_size Number of left and right indices
+* @Param max_index_value The maximum value an index can have in r_index_ptr
+* @tparam index_type The type of data associated with index_ptr
+* @tparam size_type The data type used for size calculations
+*
+* @Returns  cudaSuccess upon successful completion of append call. Otherwise returns
+* the appropriate CUDA error code
+*/
+/* ----------------------------------------------------------------------------*/
+template <typename index_type, typename size_type, typename MemAlloc = rmm_temp_allocator>
+gdf_error append_full_join_indices(
+        index_type ** l_index_ptr,
+        index_type ** r_index_ptr,
+        size_type &index_capacity,
+        size_type &index_size,
+        const size_type max_index_value,
+        MemAlloc& allocator,
+        cudaStream_t stream) {
+    gdf_error err;
+    //Get array of indices that do not appear in r_index_ptr
+    DeviceVector<index_type> unmatched_indices =
+        create_missing_indices(
+                *r_index_ptr, max_index_value, index_size, thrust::cuda::par(allocator).on(stream));
+    CUDA_CHECK_LAST()
+
+    //Expand l_index_ptr and r_index_ptr if necessary
+    size_type mismatch_index_size = unmatched_indices.size();
+    size_type l_index_capacity = index_capacity;
+    size_type r_index_capacity = index_capacity;
+    err = expand_buffer(l_index_ptr, &l_index_capacity, index_size, mismatch_index_size, stream);
+    if (GDF_SUCCESS != err) return err;
+    err = expand_buffer(r_index_ptr, &r_index_capacity, index_size, mismatch_index_size, stream);
+    if (GDF_SUCCESS != err) return err;
+
+    //Copy JoinNoneValue to l_index_ptr to denote that a match does not exist on the left
+    thrust::fill(
+            thrust::cuda::par(allocator).on(stream),
```

Comment:
Please use `rmm::exec_policy(stream)`. See https://github.com/rapidsai/cudf/blob/3350027fff3960dac269eba652bd217e497a9640/cpp/src/rmm/thrust_rmm_allocator.h#L93

```suggestion
            rmm::exec_policy(stream),
```

Then you can get rid of the allocator parameter to this function (there was no reason to have this parameter anyway, since instantiating the allocator has no cost).